_posts/2021-01-05-git-rebase-merge.md

```diff
@@ -0,0 +1,202 @@
+
+When the team chooses to go for feature-based workflow, then the git merge is the right choice because of the following reasons,
+
+- It helps in preserving the commit history, and one needs not to worry about the changing history and commits.
+- Avoids un-necessary git reverts o resets.
+- A complete feature branch can easily be re-integrated with the help of it.
+
+Contrary to this, if one wants a more linear history then the git rebase is the best option. It helps in avoiding un-necessary commits by keeping the changes linear and more centralized.
+
+One needs to be very careful while applying the rebase because if it is considered incorrectly, it can cause some serious issues.
+
+## Common Myths
+
+While it comes to rebasing and merging, most people hesitate to use git rebase as compared to merge. So we will figure out some common myths about git rebase and then compare it with merge, and we will also demonstrate when to use either one of them.
+
+So the basic purpose of Git Rebase and Git Merge is the same i.e. they help us to bring changes from other branches into ours. The following two visuals demonstrate the working of two developers while they are using git merge and git rebase respectively.
+
+![Git Merge Commit](/assets/img/posts/git-rebase-merge/git-merge-commit.png)
+
+![Git Rebase Commit](/assets/img/posts/git-rebase-merge/git-rebase-commit.png)
+
+Now we will have a look at the dangerous part about the git rebasing, it can be seen from the visual given below that it re-writes the history. So if someone else checkout your branch before we rebase yours then it would be really hard to figure out what the history of each branch is.
+
+![Git Rebase Rewrites Commit History](/assets/img/posts/git-rebase-merge/git-rebase-history.png)
+
+Just to summarize,
+
+- Rebase replays your commits on top of the new base.
+- Rebase rewrites history by creating new commits.
+- Rebase keeps the Git history clean.
+
+Some of the key points to keep in mind are,
+
+- Rebase your own local branches.
+- Don't rebase public branches – master.
+- Undo rebase with git reflog.
+
+## Conclusion
+
+Now it is time to summarize what we have discusses so far. For inconsistent repositories git rebase is not the most suitable option because the feature branch keeps on changing but for individuals rebasing provides a lot of ease. If one wants to maintain the history track, then one must go for the merging option because merging preserves the history while rebase just overwrites it.
+
+However, if we have a complex history and we want to streamline it, then interactive rebasing can be very useful. It can help us to remove undesirable commits, squash two or more commits into each other, also providing the option to commit messages. Rebase focuses on presenting one commit at a time, whereas merging focuses on presenting all at once. But it should be kept in mind that reverting rebase is much more difficult than reverting merge if there are many conflicts.

```

Comment:
We have never explained what interactive rebasing is. Perhaps add a paragraph in the "What is Rebasing?" section to explain that? I

Reply:
Added explanation of interactive rebasing in the What is Rebasing section